admin/server/reports.go

```diff
@@ -502,6 +561,154 @@ func (s *Server) generateReportName(ctx context.Context, depl *database.Deployme
 	return uuid.New().String(), nil
 }
 
+func (s *Server) issueMagicTokensForExternalEmails(ctx context.Context, orgID, projectID, reportName string, ownerID *string, opts *adminv1.ReportOptions, emails []string) (map[string]string, error) {
+	emailSet := make(map[string]struct{}, len(emails))
+	for _, email := range emails {
+		emailSet[email] = struct{}{}
+	}
+
+	emailTokens := make(map[string]string, len(emails))
+
+	// check if magic token already exists for the recipient, recipient list might have changed
+	tkns, err := s.admin.DB.FindReportTokensWithSecret(ctx, reportName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find report tokens: %w", err)
+	}
+
+	var unusedTknIDs []string
+	for _, tkn := range tkns {
+		if _, ok := emailSet[tkn.RecipientEmail]; ok {
```

Comment:
Should it maybe use `strings.ToLower` when setting/getting values from the email set?